design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,298 @@
+Support Dynamic Primary Pod NIC Name
+=
+
+# Summary
+
+This proposal proposes a mechanism to dynamically determine the primary pod interface for KuveVirt virtual machines (
+VMs).
+The goal is to move away from the hard coded `eth0` interface and allow flexible, environment-driven interface
+selection.
+
+# Overview
+
+## Motivation
+
+In KubeVirt, each VM runs as a Kubernetes pod, and these pods have network interfaces attached to different networks.
+The default primary pod interface is often named `eth0`.
+However, users might require custom network setups where the pod’s primary interface is not named `eth0`.
+Currently, KubeVirt attaches the VM’s interface destined to the pod network to the pod’s `eth0` interface.
+
+When creating a VM with secondary networks, KubeVirt specifies the pod interfaces’ names for all secondary interfaces.
+It is currently not possible to specify the primary interface’s name when templating the virt-launcher pod, since there
+is no K8s API that enables it.
+
+## Goals
+
+- Allow the primary pod interface name to be dynamically determined based on the environment.
+- Ensure backward compatibility, so existing VMs with `eth0` as the primary interface do not break.
+- Handle the scenario of a non-eth0 primary interface with a binding plugin.
+
+## Non Goals
+
+- Change Kubernetes' API.
+- Change Multus' API.
+
+## Definition of Users
+
+- Cluster Administrator
+- VM owner
+
+## User Stories
+
+- As a cluster administrator I want to use a CNI that configures pods’ primary interface to a name other than `eth0`.
+
+- As a VM owner I want my VMs to be connected to the cluster's default network, as they do today.
+
+- As a VM owner I want a correct interface status reporting.
+
+## Repos
+
+- https://github.com/kubevirt/kubevirt
+
+# Design
+
+## Using Multus’ network-status annotation
+
+When KubeVirt is used in a cluster that has Multus as its default CNI, Multus adds the
+`k8s.v1.cni.cncf.io/network-status` annotation
+to the virt-launcher pod following its network attachment process.
+
+This annotation includes information about all the network interfaces attached to the pod, for example:
+
+```json
+[
+  {
+    "name": "k8s-pod-network",
+    "ips": [
+      "10.244.196.146",
+      "fd10:244::c491"
+    ],
+    "default": true,
+    "dns": {}
+  },
+  {
+    "name": "meganet",
+    "interface": "pod7e0055a6880",
+    "mac": "8a:37:d9:e7:0f:18",
+    "dns": {}
+  }
+]
+```
+
+This annotation could be parsed to infer the default (primary) network interface instead of relying on a hard coded
+interface name (like eth0).
+
+If the network-status annotation is missing or doesn’t specify a default interface, KubeVirt will fall back to using
+`eth0` as the default interface, maintaining backward compatibility.
+
+> [!NOTE]
+> The VMI controller currently uses this annotation for:
+> - Updating the VMI status interfaces section with Multus info-source.
```

Comment:
Just a note that kubevirt is using network-status just to check interface existence but it ignores the address right now.

Reply:
Yes, currently the VMI controller ignores most of the data present in the network-status annotation.